crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,698 @@
+
+        let snapshot = mock_snapshot_metadata(
+            "/data.sst".to_owned(),
+            source_dir.to_string_lossy().to_string(),
+        );
+
+        let snapshots_destination: TempDir = TempDir::new()?;
+        let destination_dir = snapshots_destination.path().to_owned();
+        let opts = SnapshotsOptions {
+            destination: Some(
+                Url::from_file_path(snapshots_destination.path())
+                    .unwrap()
+                    .to_string(),
+            ),
+            ..SnapshotsOptions::default()
+        };
+        let repository = SnapshotRepository::create(&opts).await?.unwrap();
+
+        // Write invalid JSON to latest.json
+        let latest_path = destination_dir.join(format!("{}/latest.json", PartitionId::MIN));
+        tokio::fs::create_dir_all(latest_path.parent().unwrap()).await?;
+        info!("Creating file: {:?}", latest_path);
+        let mut latest = tokio::fs::File::create(&latest_path).await?;
+        latest.write_all(b"not valid json").await?;
+
+        assert!(repository.put(&snapshot, source_dir).await.is_err());
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_update_existing_snapshot_with_newer() -> anyhow::Result<()> {
+        tracing_subscriber::registry()
+            .with(fmt::layer())
+            .with(EnvFilter::from_default_env())
+            .init();
+
+        let snapshot_source = TempDir::new()?;
+        let source_dir = snapshot_source.path().to_path_buf();
+
+        let mut data = tokio::fs::File::create(source_dir.join("data.sst")).await?;
+        data.write_all(b"snapshot-data").await?;
+
+        let mut snapshot1 = mock_snapshot_metadata(
+            "/data.sst".to_owned(),
+            source_dir.to_string_lossy().to_string(),
+        );
+        snapshot1.min_applied_lsn = Lsn::new(
+            SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_millis() as u64,
+        );
+
+        let snapshots_destination = TempDir::new()?;
+        // #[cfg(not(feature = "s3-integration-test"))]
+        let opts = SnapshotsOptions {
+            destination: Some(
+                Url::from_file_path(snapshots_destination.path())
+                    .unwrap()
+                    .to_string(),
+            ),
+            ..SnapshotsOptions::default()
+        };
+
+        // We can't do this due to running tests with --all-features but the following
+        // code may be used to test conditional updates on S3:
+        //
+        // #[cfg(feature = "s3-integration-test")]
+        // let opts = SnapshotsOptions {
+        //     destination: Some(format!(
+        //         "s3://{}/integration-test",
+        //         std::env::var("RESTATE_S3_INTEGRATION_TEST_BUCKET_NAME")
+        //             .expect("RESTATE_S3_INTEGRATION_TEST_BUCKET_NAME must be set")
+        //     )),
+        //     ..SnapshotsOptions::default()
+        // };
```

Comment:
Could this be a separate test that we only run if `RESTATE_S3_INTEGRATION_TEST_BUCKET_NAME` has been specified and then calls into the same `update_existing_snapshot_with_newer_inner(snapshot_opts: SnapshotsOptions)`?

Reply:
Let me try it - this sounds reasonable!